src/main.rs

```diff
@@ -18,9 +19,16 @@ fn main() {
         compiler.add_file(&fname, &contents);
 
         let parser = Parser::new(compiler, span_offset);
-
         compiler = parser.parse();
     }
 
     compiler.print();
+
+    if !compiler.errors.is_empty() {
+        exit(1);
+    }
+
+    let mut resolver = Resolver::new(&compiler);
+    resolver.resolve();
+    resolver.print();
```

Comment:
so the thing I'm not sure about is after this pass we do typechecking. Do we pass them the resolver? It's be nicer if everything was in one place, like putting the resolved info back into the compiler and using that later.

Reply:
The Resolver is heavily based on the sample project you showed me, so adding typechecker into this pass should be easy (we'd rename Resolver to Typechecker). Or we'll have a separate typechecking pass, eitherr way. Even if we merge the two passes, I'd still keep the type checking and name binding APIs and data structures separate because it makes the code easier to understand, at least for me.

In the end, we need to have EngineState to make it work with the current Nushell. Maybe the way to go would be to have EngineState (Compiler) holding everything and then a separate data structure for each pass (Ast, Resolver, Typechecker) holding an immutable reference to the EngineState. After each pass, the changes would get merged into the EngineState (Compiler), similar to StateWorkingSet/StateDelta we have in Nushell, but more granular as we have more passes now.

I'm thinking that if we go this way, I could make it a separate PR, just to keep the scope of this one somewhat contained.

Reply:
I think we want to keep name resolution and typechecking separate. What I'm asking about is passing Resolver to the typechecker as a parameter.

Reply:
Yeah, with this style, we'd need to pass `&Resolver` to the `Typechecker`. But I'm not against changing it to another style, like the big EngineState/Compiler I described above. I'd rather make a separate PR for it, though, to keep the PRs cleaner.

Reply:
@sophiajt I changed it so that Resolver is now merged into the compiler after the name binding pass. Then, &Compiler can be passed to the type checker.